include/seastar/core/future.hh

```diff
@@ -1122,7 +1152,12 @@ private:
     then_wrapped_impl(Func&& func) noexcept {
         using futurator = futurize<std::result_of_t<Func(future)>>;
         if (available() && !need_preempt()) {
-            return futurator::apply(std::forward<Func>(func), future(get_available_state()));
+            if (_promise) {
+              detach_promise();
+            }
+            auto futret = futurator::apply(std::forward<Func>(func), std::move(*this));
+            _state.ignore_result();
```

Comment:
It's unlikely that somebody had taken the future and left the result inside.  We could hint the `if` inside `ignore_result()` accordingly.